wrench/src/reftest.rs

```diff
@@ -93,108 +93,144 @@ impl ReftestImage {
+        let (tx, rx) = channel();
+        wrench.renderer.set_render_notifier(Box::new(Notifier { tx: tx }));
 
-pub fn run_reftests(wrench: &mut Wrench, window: &mut WindowWrapper, filename: &str) {
-    // setup a notifier so we can wait for frames to be finished
-    struct Notifier {
-        tx: Sender<()>,
-    };
-    impl RenderNotifier for Notifier {
-        fn new_frame_ready(&mut self) {
-            self.tx.send(()).unwrap();
+        ReftestHarness {
+            wrench: wrench,
+            window: window,
+            rx: rx,
         }
-        fn new_scroll_frame_ready(&mut self, _composite_needed: bool) {}
     }
-    let (tx, rx) = channel();
-    wrench.renderer.set_render_notifier(Box::new(Notifier { tx: tx }));
 
-    let mut total_passing = 0;
-    let mut total_failing = 0;
+    pub fn run(mut self, base_manifest: &Path, specific_reftest: Option<&Path>) {
+        let manifest = ReftestManifest::new(base_manifest);
+        let mut total_passing = 0;
+        let mut total_failing = 0;
+
+        match specific_reftest {
+            Some(path) => {
+                let specific_reftest = manifest.find(path).expect("can't find reftest in manifest");
+                let success = self.run_reftest(specific_reftest);
+
+                if success {
+                    total_passing += 1;
+                } else {
+                    total_failing += 1;
+                }
+            }
+            None => {
+                for t in manifest.reftests.iter() {
+                    let success = self.run_reftest(&t);
+
+                    if success {
+                        total_passing += 1;
+                    } else {
+                        total_failing += 1;
+                    }
+                }
+            }
+        }
 
-    parse_reftests(Path::new(filename), &mut |t: Reftest| {
+        println!("REFTEST INFO | {} passing, {} failing", total_passing, total_failing);
+
+        if total_failing > 0 {
+            // panic here so that we fail CI
+            panic!();
+        }
+    }
+
+    fn run_reftest(&mut self, t: &Reftest) -> bool {
         let name = match t.op {
             ReftestOp::Equal => format!("{} == {}", t.test.display(), t.reference.display()),
             ReftestOp::NotEqual => format!("{} != {}", t.test.display(), t.reference.display()),
         };
 
         println!("REFTEST {}", name);
 
-        let test = render_yaml(wrench, window, t.test, &rx);
-        let reference = render_yaml(wrench, window, t.reference, &rx);
+        let test = self.render_yaml(t.test.as_path());
+        let reference = self.render_yaml(t.reference.as_path());
         let comparison = test.compare(&reference);
 
-        let success = match (t.op, comparison) {
-            (ReftestOp::Equal, ReftestImageComparison::Equal) => true,
-            (ReftestOp::Equal, ReftestImageComparison::NotEqual { max_difference, count_different }) => {
+        let success = match (&t.op, comparison) {
```

Comment:
t is now &Reftest and so match (t.op, comparison) is complaining on "moving out of borrowed content"